bibo/internals.py

```diff
@@ -192,3 +195,50 @@ def complete_path(ctx, args, incomplete):
         if os.access(path, os.R_OK):
             options.append((path, os.path.basename(path)))
     return options
+
+
+def match_case(substring, target):
+    """
+    Return the string `substring` with the same case as in the `target`
+    string.
+    """
+    index = target.lower().find(substring.lower())
+    if index < 0:
+        msg = "Failed to match case: {} not in {}".format(substring, target)
+        raise ValueError(msg)
+    return target[index : (index + len(substring))]
+
+
+_ansi_bold = "\033[1m"
+_ansi_unbold = "\033[22m"
+
+
+def highlight_text(text, highlight):
+    """
+    Return `text` with sub-string `highlight` in bold.
+    """
+    text = re.sub(
+        highlight, "{}{}{}".format(_ansi_bold, highlight, _ansi_unbold), text,
+    )
+    # Drop ANSI unbold followed by ANSI bold
+    return text.replace(_ansi_unbold + _ansi_bold, "")
+
+
+def highlight_match(text, result, extra_match_info=None):
```

Comment:
Docstring! I found this a tad confusing at first (before going on to look at `_recursive_search` etc). For example, knowing off the bat that `text` is a string and `result` is `models.SearchResult` which has nested results is useful.

Reply:
Agree! Will add. Was also thinking about adding type annotations and use [mypy](http://mypy-lang.org/) check in CI. What do you think?